### DISCUSSION TITLE
The type inference for resolveHref has changed.

### DISCUSSION THREAD

**Discussion by jpqp** [vercel/next.js#46813] [Help]
### Summary

As of the version upgrade to next 13.2.3, the response to resolveHref (path: next/dist/shared/lib/router/utils/resolve-href) now contains undefined, is this a specification?

↓ path becomes string | undefined
```
import { resolveHref } from 'next/dist/shared/lib/router/utils/resolve-href';
const [, path] = resolveHref(Router, urlObject, true);
```

### Additional information

```js
file next/dist/shared/lib/router/utils/resolve-href/resolveHref

Has the type inference on line 100 changed?


94  // if the origin didn't change, it means we received a relative href
95  const resolvedHref =
96    finalUrl.origin === base.origin
97      ? finalUrl.href.slice(finalUrl.origin.length)
98      : finalUrl.href
99     
100    return resolveAs 
101      ? [resolvedHref, interpolatedAs || resolvedHref]
102      : resolvedHref
```
```


### Example

_No response_

**Comment by icyJoseph** [comment]
First of all, this is what happens when we rely on internals, things move around and we just have to adapt to it. There's no spec, these are internals.

What version are you updating from? 

```ts
export function resolveHref(
  router: NextRouter,
  href: Url,
  resolveAs: true
): [string, string] | [string]
export function resolveHref(
  router: NextRouter,
  href: Url,
  resolveAs?: false
): string
export function resolveHref(
  router: NextRouter,
  href: Url,
  resolveAs?: boolean
): [string, string] | [string] | string
```

This function is overloaded, which means that the inferred return type depends on its inputs, but most importantly, we do see, that you have to check that you have a tuple of two elements.

That is you have to check now that in:

```ts
const [, path] = resolveHref(Router, urlObject, true);
```

`path` can be undefined, even more, the result type might not even be an array!

### Detective hat on

What I see is that 3 years ago the `resolveHref` function looked like, https://github.com/vercel/next.js/blame/f012eab00f9abd2b34f82edc42d84458cbf414a8/packages/next/shared/lib/router/router.ts:

```ts
export function resolveHref(
  router: NextRouter,
  href: Url,
  resolveAs?: boolean
): string {
```

And was used like:

```ts
  let [resolvedHref, resolvedAs] = resolveHref(router, url, true)
```

Which is really strange, so I had to look into the implementation (from 3 years ago) and found this:

```ts
  return (
      resolveAs ? [resolvedHref, interpolatedAs || resolvedHref] : resolvedHref
    ) as string
```

Which means that the tuple return types had been casted to strings, not good.

Anyway, not sure what you are looking for here, you shouldn't rely on internals, if they changed, you have to adapt.

And we did learn that, 2-3 years ago, there was some TypeScript trickery going on around with these lines of code.
